articles/cross_compilation_build_tools_design.md

```diff
@@ -0,0 +1,248 @@
+  file is already available, for example when new dependencies are added to some
+  package in the workspace.
+- _How can we run the resulting binaries?_ Packaging the resulting binaries is
+  outside of the scope of this document. The cross-compiled binaries will be
+  available in the workspace is a directory corresponding to the architecture-OS
+  pair, e.g. `armhf-ubuntu_bionic`. We make the assumption that the user will
+  copy the whole workspace to the remote device (including src/) and run `rosdep`
+  on the target platform, to install the dependencies before running the binaries.
+
+### Maintenance and testing
+
+Regarding maintenance, in order to support cross-compilation for a new platform
+we would need to:
+
+- Prepare and publish a base ROS 2 Docker image for the new platform, both
+  fetching the dependencies using the ROS 2 source using `rosdep install`, and
+  from the pre-built binary if available for that platform.
+- Adapt the toolchain file for the new platform, if required.
+- Extend the `cc-build` plugin to include dependencies for the compiler for
+  the new platform, and any modifications needed to use a suitable toolchain
+  file for the platform.
+
+We could then test cross-compile support for the new platform by building the
+base  ROS 2 Docker image, and a custom sysroot image for some reference package
+(e.g. [ApexAI/performance_test](https://github.com/ApexAI/performance_test/)),
+and then launching a Docker container for the that sysroot image, running
+`colcon test` using the cross-compiled binaries. The container would be
+launched with a [bind mount](https://docs.docker.com/storage/bind-mounts/) of the
+workspace so they have access to the compiled binaries.
+
+In order to simplify the development of base ROS 2 Docker images for new
+architecture-OS combinations, on a second iteration of this plugin we might add
+a new verb `colcon cc-build-sysroot-base-image` that will 1) copy the required
+qemu files and ROS 2 source into a temporary file; 2) build a Docker image from
+a specified Dockerfile, with access to the those resources; 3) optionally
+publish the image into a Docker registry, using a suitable naming convention like
+`${docker_repo_uri}:${os}_${arch}_${prebuilt?}_${rosdistro}`. Dockerfiles
+compatible with this command, for each architecture-OS pair would be keep on the
+`ros2/cross_compile` repository, so they can be used in CI/CI pipeline if needed.
+
+### Advantages of the approach
+
+The proposed design leads to a simple development experience, where cross
+compilation is triggered by a one-liner command.
+The design is extensible and new platforms can be supported easily.
+
+For simple packages with additional dependencies, there is no need to build the
+sysroot locally with Docker, and a readily available base Docker image can be
+used. For packages with custom dependencies, the base sysroot can be extended
+using `rosdep install`.
```

Comment:
I'd clarify that `rosdep install` here needs to be run inside QEmu.

Reply:
(through `binfmt_misc`)

Reply:
I'll add explicit reference to the QEmu usage there